sgl-kernel/csrc/cpu/mamba/fla.cpp

```diff
@@ -0,0 +1,1153 @@
+            /* A */ kg_transpose,
+            /* B */ v_prime_pack_reduced,
+            /* C */ kgv);
+        // last_recurrent_state = 1) + 2)
+        for (int64_t m = 0; m < qk_head_size; m++) {
+          at::vec::map2<float>(
+              [](at::vec::Vectorized<float> x, at::vec::Vectorized<float> y) { return x + y; },
+              curr_last_recurrent_state + m * v_head_size,
+              curr_last_recurrent_state + m * v_head_size,
+              kgv + m * v_head_size,
+              v_head_size);
+        }
+      }
+
+      // core_attn_out -> output
+      // output: [B, T, HV, EV]
+      // core_attn_out: [B, HV, padded_T, EV]
+      auto curr_out = out_ptr + h * oStrideH;
+      for (int64_t m = 0; m < seq_len; m++) {
+        at::vec::map<scalar_t>(
+            [](at::vec::Vectorized<float> x) { return x; },
+            curr_out + m * oStrideT,
+            curr_core_attn_out + m * v_head_size,
+            v_head_size);
+      }
+
+      // Move to the next query
+      data_index_step(b, batch_size, h, v_num_head);
+    }
+  });
+}
+}  // anonymous namespace
+
+// query: [B, T, HK, EK]
+// key: [B, T, HK, EK]
+// value: [B, T, HV, EV]
+// g: [B, T, HV] FP32
+// beta: [B, T, HV]
+// initial_state: [N, HV, EK, EV] FP32
+// output_final_state: bool
+// cu_seqlens: [N + 1] INT32
+// head_first: bool
+// use_qk_l2norm_in_kernel: bool
+std::tuple<at::Tensor, at::Tensor> chunk_gated_delta_rule_cpu(
+    at::Tensor& query,
+    at::Tensor& key,
+    at::Tensor& value,
+    at::Tensor& g,
+    at::Tensor& beta,
+    at::Tensor& initial_state,
+    bool output_final_state,
+    at::Tensor& cu_seqlens,
+    bool head_first,
+    bool use_qk_l2norm_in_kernel) {
+  RECORD_FUNCTION(
+      "sgl-kernel::chunk_gated_delta_rule_cpu", std::vector<c10::IValue>({query, key, value, g, beta, initial_state}));
+
+  TORCH_CHECK(head_first == false, "chunk_gated_delta_rule_cpu does not support head first");
+  TORCH_CHECK(
+      query.dtype() == at::kBFloat16 && query.dtype() == key.dtype() && query.dtype() == value.dtype() &&
+      query.dtype() == beta.dtype());
+  TORCH_CHECK(g.dtype() == at::kFloat && g.dtype() == initial_state.dtype());
+  TORCH_CHECK(cu_seqlens.dtype() == at::kInt);
+  CHECK_DIM(4, query);
+  CHECK_DIM(4, key);
+  CHECK_DIM(4, value);
+  CHECK_DIM(3, g);
+  CHECK_DIM(3, beta);
+  CHECK_DIM(1, cu_seqlens);
+  CHECK_DIM(4, initial_state);
+  CHECK_LAST_DIM_CONTIGUOUS_INPUT(query);
+  CHECK_LAST_DIM_CONTIGUOUS_INPUT(key);
+  CHECK_LAST_DIM_CONTIGUOUS_INPUT(value);
+  CHECK_CONTIGUOUS(g);
+  CHECK_CONTIGUOUS(beta);
+  CHECK_CONTIGUOUS(initial_state);
+  int64_t B = query.size(0);
+  int64_t T = query.size(1);
+  int64_t HK = query.size(2);
+  int64_t EK = query.size(3);
+  int64_t HV = value.size(2);
+  int64_t EV = value.size(3);
+  CHECK_EQ(B, 1);
+  CHECK_EQ(key.size(0), B);
+  CHECK_EQ(key.size(1), T);
+  CHECK_EQ(key.size(2), HK);
+  CHECK_EQ(key.size(3), EK);
+  CHECK_EQ(value.size(0), B);
```

Comment:
this function has many input tensors, we can simplify these checks by having a helper function, say:
```
template <bool is_last_dim_contiguous>
static inline  void CHECK_INPUT_SHAPE_DTYPE(at::Tensor& tensor, const std::vector<int64_t>& sizes, at::ScalarType st) {...}

}
```